File: implementations/Python/03_Fibonacci/01_default/main.py
#!/usr/bin/env python

from __future__ import print_function

import sys
import time


class TimedContext(object):

    def __enter__(self):
        self.t1 = time.time()

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.t2 = time.time()
        print(self.t2 - self.t1)


def fibonacci_naive(N):
    if N < 2:
        return N
    else:
        return fibonacci_naive(N-1) + fibonacci_naive(N-2)


def fibonacci_tailrec(N, a=0, b=1):
    if N == 0:
        return a
    else:
        return fibonacci_tailrec(N-1, b, a+b)


def fibonacci_iterative(N):
    a = 0
    b = 1
    while N > 0:
        a, b = b, a + b
        N -= 1
    return a


if len(sys.argv) != 3:
    sys.exit(1)
N = int(sys.argv[1])
M = int(sys.argv[2])


# Note: Crashes without explicitly increasing recursion limit
sys.setrecursionlimit(N*2)


with TimedContext():
    f1 = fibonacci_naive(N)

with TimedContext():
    checksum_f2 = 0
    for i in xrange(M):
        checksum_f2 += fibonacci_tailrec(N)
        checksum_f2 %= 2147483647

with TimedContext():
    checksum_f3 = 0
    for i in xrange(M):
        checksum_f3 += fibonacci_iterative(N)
        checksum_f3 %= 2147483647

print(f1)
print(checksum_f2)
print(checksum_f3)

File: implementations/Scala/03_Fibonacci/01_default/src/main/scala/main.scala

object Main {

  def runTimed[T](body: => T): T = {
    val t1 = System.nanoTime()
    val res = body
    val t2 = System.nanoTime()
    println((t2 - t1).toDouble / 1000000000)
    res
  }

  def fibonacciNaive(N: Int): Long = {
    if (N < 2) {
      N
    } else {
      fibonacciNaive(N-1) + fibonacciNaive(N-2)
    }
  }

  def fibonacciTailRec(N: Int, a: Long = 0, b: Long = 1): Long = {
    if (N == 0) {
      a
    } else {
      fibonacciTailRec(N-1, b, a+b)
    }
  }

  def fibonacciIterative(N: Int): Long = {
    var a = 0L
    var b = 1L
    var Ncopy = N
    while (Ncopy > 0) {
      val tmp = a
      a = b
      b = b + tmp
      Ncopy -= 1
    }
    a
  }

  def main(args: Array[String]): Unit = {

    if (args.length != 2) {
      System.exit(1)
    }
    val N = args(0).toInt
    val M = args(0).toInt

    val f1 = runTimed{
      fibonacciNaive(N)
    }

    val checksumF2 = runTimed{
      var checksum = 0L
      for (i <- 0 until M) {
        checksum += fibonacciTailRec(N)
        checksum %= 2147483647
      }
      checksum
    }

    val checksumF3 = runTimed{
      var checksum = 0L
      for (i <- 0 until M) {
        checksum += fibonacciIterative(N)
        checksum %= 2147483647
      }
      checksum
    }

    println(f1)
    println(checksumF2)
    println(checksumF3)
  }

}